go-controller/pkg/ovn/address_set/address_set.go

```diff
@@ -53,16 +52,16 @@ type AddressSet interface {
 	// GetName returns the descriptive name of the address set
 	GetName() string
 	// AddIPs adds the array of IPs to the address set
-	AddIPs(ip []net.IP) error
+	AddIPs(ips []string) error
 	// AddIPsReturnOps returns the ops needed to add the array of IPs to the address set
-	AddIPsReturnOps(ip []net.IP) ([]ovsdb.Operation, error)
+	AddIPsReturnOps(ips []string) ([]ovsdb.Operation, error)
 	// GetIPs gets the list of v4 & v6 IPs from the address set
 	GetIPs() ([]string, []string)
 	// SetIPs sets the address set to the given array of addresses
-	SetIPs(ip []net.IP) error
-	DeleteIPs(ip []net.IP) error
+	SetIPs(ips []string) error
+	DeleteIPs(ips []string) error
 	// DeleteIPsReturnOps returns the ops needed to delete the array of IPs from the address set
-	DeleteIPsReturnOps(ip []net.IP) ([]ovsdb.Operation, error)
+	DeleteIPsReturnOps(ips []string) ([]ovsdb.Operation, error)
```

Comment:
> The interface still mentions IPs instead of addresses. Like for example, `AddIPs` vs `AddAddresses`. The whole address set implementation shoud be mindful that these are no longer just IPs so basically the 
mention of IPs should be gone.

yeah this is a great point!

> 
> Maybe you don't plan to commit to non IP addresses yet, like MACs. In that case, should we validate somewhere that these are indeed IPs or CIDRs?

correct I don't commit to things like MACs yet.. so far in all the places of our code we have been using only IPs and now from ANP we use CIDRs... and before passing this into this function before since it was net.IP most of the places did the parseIP/parseCIDR got the ip and passed it here, so if we do add a validation inside the address set util I should go and cleanup all places to ensure we don't double parse things?.. but I agree some sort of validation is perhaps needed now that we have changed this to a string type, but I am not sure how to go about the problem around "input was net.IP" now we made it "string" and then now we need to validate it again!

> 
> However I feel we might incur in a performance cost. Because now we are converting IPs to strings, then validating that those strings are actually IPs (not only potentially just here, but I see that pattern emerging also around other parts of the code because of this change). 

ah yes exactly this is what I meant ++ correct.

> So perhaps it would make sense to have dedicated methods, like
> 
> ```
> AddIPs(ips []net.IP)
> AddIPNets(ipnets []net.IPNet)
> AddMacs(macs []net.HardwareAddress)
> AddAddresses(addresses []string) <-- in this case we woudl be in charge of validating/splitting
> ```

yeah.. maybe that was the right thing to do from start... ugh how about I go back to not changing anything for AddIPs interfaces, leave them as is.. and simply add the AddIPNets for now which gets me the cidr bits?
someone else in the future introducing macs and other types can add on top of that ? I have some cycles now so let me see if I can spend some hours on this cleanup

> 
> That way we potentially provide the user they most convinient API to do the conversions and validations that they see fit when they see fit.
> 
> However this might be getting it too far. So if you think we are fine on this last point, then lets let it be for now. Do let me now what your plans with the naming & validation.

let me try it and see how we do!

Reply:
@jcaamano and I tried this out offline, and discussed about this, I want to summarize the reasoning why we don't want to go head with the above suggestion so that we remember and its historically tracked for future purposes.

```
the problem with using different types of utils for each type example:
AddIPs(ips []net.IP)
AddIPNets(ipnets []net.IPNet)
AddMacs(macs []net.HardwareAddress)
```

is that each time we want to mutate the same field ; example in case of ANP we will need to do AddIPsReturnOps and AddIPNetsReturnOps calls separately this is going to cost us on performance and churn which we think is more than the perf hit of parser validations at this stage.

SO instead of that approach we went with trying to use generics to solve this problem, something like:
```
type AddressSetTypes interface {
	net.IP | net.IPNet
	String() string
}
```
```
type AddressSetFactory[T AddressSetTypes] interface {
    // NewAddressSetOps returns a new object that implements AddressSet
	// and contains the given IPs, or an error. Internally it creates
	// ops to create an address set for IPv4 and IPv6 each.
	NewAddressSetOps(dbIDs *libovsdbops.DbObjectIDs, ips []T) (AddressSet[T], []ovsdb.Operation, error)
}
```
but now at both set and get methods of the factory utils we need a way to pass both types net.IP and net.IPnet which is tricky and couldn't find a way to do that without having to parse those types differently.
doing 
```
type AddressSetFactory interface {
    // NewAddressSetOps returns a new object that implements AddressSet
	// and contains the given IPs, or an error. Internally it creates
	// ops to create an address set for IPv4 and IPv6 each.
	NewAddressSetOps(dbIDs *libovsdbops.DbObjectIDs, ips AddressSetTypes) (AddressSet, []ovsdb.Operation, error)
}
```
without generics also doesn't work since we end up with `cannot use type AddressSetTypes outside a type constraint: interface contains type constraints`

so as of now since OVN's address set table is itself an array of strings, we are going ahead with just using strings here in the util.

But this means we need to take note to add some developer docs around this because people need to be careful around validating what they send and also validating what they receive from OVS DB server